### DISCUSSION TITLE
General questions regarding your implementation, comparison to HGS-CVRP (Vidal 2022)

### DISCUSSION THREAD

**Discussion by bmorlo** [PyVRP/PyVRP#600] [Q&A]
Hi,

after finishing Vidal‘s papers and going through your 2023 paper once more I noticed a few differences. Allow me to also ask some more general questions pertaining to both implementations.

- How is the subtour elimination taken care of? It is something that I have been asking myself for a very long time.
- You are using SREX instead of ordered crossover (OX). Why is that? Do I understand it correctly that no SPLIT algorithm is needed since SREX considers parent _routes_ separately instead of parent _solutions_ as concatenated mega-routes?
- Did I understand it correctly that your local search implementation is very similar to the one in HGS-CVRP except that you have not implemented all the operators from his 2012 paper? (Apologies if you have)
- Where do you think does the difference in performance, as reported in your 2023 paper, between PyVRP and HGS-CVRP mainly comes from?

Thanks and cheers as always.

**Comment by N-Wouda** [comment]
> * How is the subtour elimination taken care of? It is something that I have been asking myself for a very long time.

PyVRP is a heuristic, so we don't have/need this.

> * You are using SREX instead of ordered crossover (OX). Why is that? Do I understand it correctly that no SPLIT algorithm is needed since SREX considers parent _routes_ separately instead of parent _solutions_ as concatenated mega-routes?

Because it's better and obviates the need for the SPLIT algorithm.

> * Did I understand it correctly that your local search implementation is very similar to the one in HGS-CVRP except that you have not implemented all the operators from his 2012 paper? (Apologies if you have)

We removed all those we appear to be 'missing' since they were not effective enough.

> * Where do you think does the difference in performance, as reported in your 2023 paper, between PyVRP and HGS-CVRP mainly comes from?

Simplification, and a more general focus. We could probably get competitive with HGS-CVRP but it just isn't worth it to us in terms of maintenance and tuning effort.

**Reply by bmorlo** [reply]
> PyVRP is a heuristic, so we don't have/need this.

Got it. Makes sense. But then again: You consider the e.g. excessLoad penalties to penalise infeasible solutions. Shouldn't the same be done with other (subtour) infeasible solutions? Maybe I have not yet understood this properly.

And thanks for the other answers to my questions. They all helped!

**Reply by bmorlo** [reply]
> But then again: You consider the e.g. excessLoad penalties to penalise infeasible solutions. Shouldn't the same be done with other "subtour-infeasible" solutions?

Apologies if it sounded like the discussion was resolved. Could you point me in the right direction here? I.e. how do we make sure that all feasible solutions would indeed not violate a subtour elimination constraint? It is hard for me to comprehend this since we do not actively specify this in the feasibility evaluations (e.g. !hasExcessLoad && etc.).

Many thanks.

**Reply by N-Wouda** [reply]
Required clients are inserted in the local search if they are missing from the solution.

**Reply by bmorlo** [reply]
Right. So basically, we would always have all required clients in a solution (if not they would be inserted immediately). And since a solution is basically a set of "client vectors" (aka routes) there is no need to specifically check for subtours. Did I get that correctly?

Last question:
If a solution does not contain all required clients, would it be treated as an infeasible one?

Edit: Another question. I saw in the code that the initial infeasibility penalties for `excessLoad` are 20. I also know/saw that we increase the load penalty whenever there are too many infeasible solutions. Now, **wouldn't it make sense to set the initial load penalty to something like the average distance in the instance? Wouldn't this greatly enhance the (starting) performance if the solver is used with very high distance values** (e.g. 5000) in the distance matrix? Wouldn't this skip the process of trying to increase the penalty? I tested this, but the performance seemes fine. So, how is this done?

Thanks you guys so much as always ✌️

**Reply by bmorlo** [reply]
Would be great to hear back on that as I'm going to present my work at university this week. 🙏

**Reply by N-Wouda** [reply]
I am sure that you can answer these questions yourself with some experimentation and reading the relevant bits of the code. Good luck with your presentation!

**Reply by bmorlo** [reply]
Apologies if it sounded like I did not try to find the answers in the code myself. Apologies also if it came across impolitely. However, I can assure you I tried my best, already.

I really don't want to take too much of your time. Allow me to rephrase my questions, such that you could answer them super quickly. Would help tons.

1) Do we consider a solution infeasible, if it does not contain all customers? Or can this simply not happen.

2) Could you kindly point me to the code section where the penalty terms are adjusted in the beginning such that they are at level with e.g. the average distance between clients? Or is this not done specifically.

Again, if it's not too much of an ask, your answers would be greatly appreciated.
Cheers.
